[BugFix] sys.fe_memory_usage / sys.fe_locks return misleading 'BE:{id}, host: unknown' when caller lacks OPERATE privilege

<!-- StarRocks bug report -->

### Steps to reproduce the behavior (Required)

On any shared-data cluster (FE + CN, no BE), as a user **without** the `OPERATE ON SYSTEM` privilege (i.e. anyone whose active roles don't include `db_admin` / `security_admin` / `root`):

```sql
-- Both fail the same way
SELECT * FROM sys.fe_memory_usage;
SELECT * FROM sys.fe_locks;
```

### Expected behavior (Required)

The client should receive an `Access denied` error that names the missing privilege, the same way every other `OPERATE`-gated statement does (e.g. `ADMIN SHOW CONFIG` returns `Access denied; you need (at least one of) the GRANT privilege(s) on SYSTEM ...`).

### Real behavior (Required)

The client gets a misleading "node lookup" style error that has nothing to do with the actual cause:

```
ERROR 1064 (HY000): FE RPC failure,
  address=TNetworkAddress(hostname=<fe-host>, port=9020),
  reason=Internal error processing listFeMemoryUsage: BE:10026, host: unknown
```

```
ERROR 1064 (HY000): FE RPC failure,
  address=TNetworkAddress(hostname=<fe-host>, port=9020),
  reason=Internal error processing listFeLocks: BE:10030, host: unknown
```

The `BE:<id>` segment looks like a backend-id lookup failure, and `host: unknown` reinforces that misreading. **Neither is true.** `BE:<id>` is just the CN's own id stamped on a failed pipeline-driver status, and `host: unknown` is a hard-coded literal in `DefaultCoordinator`. The actual cause is an empty-message `AccessDeniedException` that gets swallowed by three layers of error wrapping.

### Server-side stack trace (from `fe.log` of the FE that served the call)

```
ERROR  Internal error processing listFeMemoryUsage
org.apache.thrift.TException
    at com.starrocks.catalog.system.sys.SysFeMemoryUsage.listFeMemoryUsage(SysFeMemoryUsage.java:67)
    at com.starrocks.service.FrontendServiceImpl.listFeMemoryUsage(FrontendServiceImpl.java:662)
    at com.starrocks.thrift.FrontendService$Processor$listFeMemoryUsage.getResult(FrontendService.java:7041)
    at com.starrocks.thrift.FrontendService$Processor$listFeMemoryUsage.getResult(FrontendService.java:7018)
    at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:40)
    at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:40)
    at com.starrocks.common.SRTThreadPoolServer$WorkerProcess.run(SRTThreadPoolServer.java:313)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: com.starrocks.authorization.AccessDeniedException
    at com.starrocks.authorization.NativeAccessController.checkObjectTypeAction(NativeAccessController.java:321)
    at com.starrocks.authorization.NativeAccessController.checkSystemAction(NativeAccessController.java:44)
    at com.starrocks.sql.analyzer.Authorizer.checkSystemAction(Authorizer.java:66)
    at com.starrocks.catalog.system.sys.SysFeMemoryUsage.listFeMemoryUsage(SysFeMemoryUsage.java:65)
    ... 9 more
```

The coordinator wrap on the same FE:

```
com.starrocks.rpc.RpcException: FE RPC failure,
  address=TNetworkAddress(hostname=<fe-host>, port=9020),
  reason=Internal error processing listFeMemoryUsage: BE:10001, host: unknown
    at com.starrocks.qe.DefaultCoordinator.dealStatusToTryRetry(DefaultCoordinator.java:938)
    at com.starrocks.qe.DefaultCoordinator.getNext(DefaultCoordinator.java:1011)
    at com.starrocks.qe.StmtExecutor.handleQueryStmt(StmtExecutor.java:1686)
    at com.starrocks.qe.StmtExecutor.execute(StmtExecutor.java:879)
    ...
```

### Root cause

Three layers each drop information:

1. **`fe/fe-core/src/main/java/com/starrocks/authorization/NativeAccessController.java:321`** (and every other throw site in the same file) does `throw new AccessDeniedException();` — no message.
2. **`fe/fe-core/src/main/java/com/starrocks/catalog/system/sys/SysFeMemoryUsage.java:58-60`** catches the empty-message exception and rethrows as `TException(e.getMessage(), e)` — `e.getMessage()` is `null`, so the rethrown TException carries no information either. Same anti-pattern in `SysFeLocks.java:86-88`.
3. Apache Thrift then wraps the empty-message TException as the generic `TApplicationException("Internal error processing listFeMemoryUsage")`. The BE's `thrift_rpc_helper.cpp:86` adds the `FE RPC failure, address=..., reason=` prefix, `pipeline_driver_executor.cpp:193` appends `: BE:{be_id}`, and finally `DefaultCoordinator.java:947` wraps with `RpcException("unknown", status.getErrorMsg())` — that literal `"unknown"` is what produces `, host: unknown` at the very end.

So the user-facing error is essentially `[decorative-be-tag]:[decorative-host-tag]` with the real reason completely gone.

### Why this matters

- `sys.fe_memory_usage` and `sys.fe_locks` are the two main observability tables for FE memory and lock contention. They are exactly the tables you want to query during an incident, when you want clear errors, not a maze of misleading wrappers.
- Operators reading this error reasonably suspect a shared-data CN/BE node-lookup bug, lock-manager corruption, or RPC routing problem. None of those are happening — but ruling them out takes hours of code spelunking.
- The same anti-pattern (`catch (AccessDeniedException e) { throw new TException(e.getMessage(), e); }`) appears in several other system-table handlers (e.g. `SysObjectDependencies`, `FeThreadsSystemTable`), so this same opaque error surfaces from those tables too.

### Suggested direction (for the team to evaluate)

The fix likely belongs in the `catch (AccessDeniedException e)` block in `SysFeMemoryUsage.java:58-60` and `SysFeLocks.java:86-88` — it should surface a meaningful "access denied" message instead of the empty `e.getMessage()`. The deeper fix is to give `new AccessDeniedException()` a default message at every throw site in `NativeAccessController.java`, since the same null-message swallow can leak through any handler that uses the same `catch`/rethrow pattern. Happy to leave the exact shape to the maintainers.

### Workaround (for affected users until a patched release)

Grant `OPERATE ON SYSTEM` — either narrowly to the specific user, or to a role they hold:

```sql
GRANT OPERATE ON SYSTEM TO USER '<user>'@'<host>';
-- or
GRANT db_admin TO ROLE <your-observability-role>;
```

Note: `OPERATE ON SYSTEM` is a broad admin privilege that also allows `ADMIN SET CONFIG`, `ADMIN SET REPLICA STATUS`, `EXECUTE SCRIPT`, etc. There is no fine-grained "read-only system tables" privilege today. Splitting one out (e.g. `INSPECT`) so that `sys.fe_*` can be queried without granting cluster-write powers would be a useful follow-up.

### StarRocks version (Required)

```
4.0.10-5b36164
```

(Shared-data deployment, 3 FE + 30 CN.)

--- a/fe/fe-core/src/main/java/com/starrocks/catalog/system/sys/SysFeLocks.java
+++ b/fe/fe-core/src/main/java/com/starrocks/catalog/system/sys/SysFeLocks.java
@@ -19,10 +19,12 @@ import com.google.common.base.Joiner;
 import com.google.gson.JsonObject;
 import com.starrocks.authentication.UserIdentityUtils;
 import com.starrocks.authorization.AccessDeniedException;
+import com.starrocks.authorization.ObjectType;
 import com.starrocks.authorization.PrivilegeType;
 import com.starrocks.catalog.Table;
 import com.starrocks.catalog.system.SystemId;
 import com.starrocks.catalog.system.SystemTable;
+import com.starrocks.common.ErrorCode;
 import com.starrocks.common.util.concurrent.lock.LockHolder;
 import com.starrocks.common.util.concurrent.lock.LockInfo;
 import com.starrocks.common.util.concurrent.lock.LockManager;
@@ -75,7 +77,8 @@ public class SysFeLocks {
                 Authorizer.checkSystemAction(context, PrivilegeType.OPERATE);
             }
         } catch (AccessDeniedException e) {
-            throw new TException(e.getMessage(), e);
+            throw new TException(ErrorCode.ERR_ACCESS_DENIED_FOR_EXTERNAL_ACCESS_CONTROLLER.formatErrorMsg(
+                    PrivilegeType.OPERATE.name(), ObjectType.SYSTEM.name(), ""), e);
         }
 
         TFeLocksRes response = new TFeLocksRes();

--- a/fe/fe-core/src/main/java/com/starrocks/catalog/system/sys/SysFeLocks.java
+++ b/fe/fe-core/src/main/java/com/starrocks/catalog/system/sys/SysFeLocks.java
@@ -21,10 +21,10 @@ import com.starrocks.authentication.UserIdentityUtils;
 import com.starrocks.authorization.AccessDeniedException;
 import com.starrocks.authorization.ObjectType;
 import com.starrocks.authorization.PrivilegeType;
+import com.starrocks.catalog.InternalCatalog;
 import com.starrocks.catalog.Table;
 import com.starrocks.catalog.system.SystemId;
 import com.starrocks.catalog.system.SystemTable;
-import com.starrocks.common.ErrorCode;
 import com.starrocks.common.util.concurrent.lock.LockHolder;
 import com.starrocks.common.util.concurrent.lock.LockInfo;
 import com.starrocks.common.util.concurrent.lock.LockManager;
@@ -77,8 +77,9 @@ public class SysFeLocks {
                 Authorizer.checkSystemAction(context, PrivilegeType.OPERATE);
             }
         } catch (AccessDeniedException e) {
-            throw new TException(ErrorCode.ERR_ACCESS_DENIED_FOR_EXTERNAL_ACCESS_CONTROLLER.formatErrorMsg(
-                    PrivilegeType.OPERATE.name(), ObjectType.SYSTEM.name(), ""), e);
+            AccessDeniedException.reportAccessDenied(InternalCatalog.DEFAULT_INTERNAL_CATALOG_NAME,
+                    context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
+                    PrivilegeType.OPERATE.name(), ObjectType.SYSTEM.name(), null);
         }
 
         TFeLocksRes response = new TFeLocksRes();

--- a/fe/fe-core/src/main/java/com/starrocks/catalog/system/sys/SysFeMemoryUsage.java
+++ b/fe/fe-core/src/main/java/com/starrocks/catalog/system/sys/SysFeMemoryUsage.java
@@ -18,10 +18,10 @@ import com.starrocks.authentication.UserIdentityUtils;
 import com.starrocks.authorization.AccessDeniedException;
 import com.starrocks.authorization.ObjectType;
 import com.starrocks.authorization.PrivilegeType;
+import com.starrocks.catalog.InternalCatalog;
 import com.starrocks.catalog.Table;
 import com.starrocks.catalog.system.SystemId;
 import com.starrocks.catalog.system.SystemTable;
-import com.starrocks.common.ErrorCode;
 import com.starrocks.memory.MemoryUsageTracker;
 import com.starrocks.qe.ConnectContext;
 import com.starrocks.sql.analyzer.Authorizer;
@@ -58,8 +58,9 @@ public class SysFeMemoryUsage {
         try {
             Authorizer.checkSystemAction(context, PrivilegeType.OPERATE);
         } catch (AccessDeniedException e) {
-            throw new TException(ErrorCode.ERR_ACCESS_DENIED_FOR_EXTERNAL_ACCESS_CONTROLLER.formatErrorMsg(
-                    PrivilegeType.OPERATE.name(), ObjectType.SYSTEM.name(), ""), e);
+            AccessDeniedException.reportAccessDenied(InternalCatalog.DEFAULT_INTERNAL_CATALOG_NAME,
+                    context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
+                    PrivilegeType.OPERATE.name(), ObjectType.SYSTEM.name(), null);
         }
 
         TFeMemoryRes response = new TFeMemoryRes();

--- a/fe/fe-core/src/main/java/com/starrocks/catalog/system/sys/SysFeMemoryUsage.java
+++ b/fe/fe-core/src/main/java/com/starrocks/catalog/system/sys/SysFeMemoryUsage.java
@@ -16,10 +16,12 @@ package com.starrocks.catalog.system.sys;
 
 import com.starrocks.authentication.UserIdentityUtils;
 import com.starrocks.authorization.AccessDeniedException;
+import com.starrocks.authorization.ObjectType;
 import com.starrocks.authorization.PrivilegeType;
 import com.starrocks.catalog.Table;
 import com.starrocks.catalog.system.SystemId;
 import com.starrocks.catalog.system.SystemTable;
+import com.starrocks.common.ErrorCode;
 import com.starrocks.memory.MemoryUsageTracker;
 import com.starrocks.qe.ConnectContext;
 import com.starrocks.sql.analyzer.Authorizer;
@@ -56,7 +58,8 @@ public class SysFeMemoryUsage {
         try {
             Authorizer.checkSystemAction(context, PrivilegeType.OPERATE);
         } catch (AccessDeniedException e) {
-            throw new TException(e.getMessage(), e);
+            throw new TException(ErrorCode.ERR_ACCESS_DENIED_FOR_EXTERNAL_ACCESS_CONTROLLER.formatErrorMsg(
+                    PrivilegeType.OPERATE.name(), ObjectType.SYSTEM.name(), ""), e);
         }
 
         TFeMemoryRes response = new TFeMemoryRes();

--- a/fe/fe-core/src/test/java/com/starrocks/catalog/system/sys/SysFeLocksTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/catalog/system/sys/SysFeLocksTest.java
@@ -60,11 +60,11 @@ public class SysFeLocksTest {
             }
         };
 
-        TException ex = assertThrows(TException.class, () -> SysFeLocks.listLocks(req, true));
+        Exception ex = assertThrows(Exception.class, () -> SysFeLocks.listLocks(req, true));
         assertNotNull(ex.getMessage(), "AccessDenied message should not be null");
-        assertTrue(ex.getMessage().contains("Access denied"), "Should match canonical format: " + ex.getMessage());
-        assertTrue(ex.getMessage().contains("OPERATE"), "Should mention OPERATE: " + ex.getMessage());
-        assertTrue(ex.getMessage().contains("SYSTEM"), "Should mention SYSTEM: " + ex.getMessage());
+        assertTrue(ex.getMessage().contains(
+                        "Access denied; you need (at least one of) the OPERATE privilege(s) on SYSTEM for this operation."),
+                "Should match canonical format: " + ex.getMessage());
     }
 
 }

--- a/fe/fe-core/src/test/java/com/starrocks/catalog/system/sys/SysFeLocksTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/catalog/system/sys/SysFeLocksTest.java
@@ -14,12 +14,20 @@
 
 package com.starrocks.catalog.system.sys;
 
+import com.starrocks.authorization.AccessDeniedException;
+import com.starrocks.authorization.PrivilegeType;
+import com.starrocks.qe.ConnectContext;
+import com.starrocks.sql.analyzer.Authorizer;
 import com.starrocks.thrift.TAuthInfo;
 import com.starrocks.thrift.TFeLocksReq;
+import mockit.Mock;
+import mockit.MockUp;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.thrift.TException;
 import org.junit.jupiter.api.Test;
 
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 public class SysFeLocksTest {
@@ -34,6 +42,29 @@ public class SysFeLocksTest {
 
         var res = SysFeLocks.listLocks(req, false);
         assertTrue(StringUtils.isNotEmpty(res.toString()));
+    }
+
+    @Test
+    public void testListLocksAccessDeniedSurfacesMessage() {
+        TFeLocksReq req = new TFeLocksReq();
+        TAuthInfo auth = new TAuthInfo();
+        auth.setUser("nopriv");
+        auth.setUser_ip("127.0.0.1");
+        req.setAuth_info(auth);
+
+        new MockUp<Authorizer>() {
+            @Mock
+            public void checkSystemAction(ConnectContext context, PrivilegeType privilegeType)
+                    throws AccessDeniedException {
+                throw new AccessDeniedException();
+            }
+        };
+
+        TException ex = assertThrows(TException.class, () -> SysFeLocks.listLocks(req, true));
+        assertNotNull(ex.getMessage(), "AccessDenied message should not be null");
+        assertTrue(ex.getMessage().contains("Access denied"), "Should match canonical format: " + ex.getMessage());
+        assertTrue(ex.getMessage().contains("OPERATE"), "Should mention OPERATE: " + ex.getMessage());
+        assertTrue(ex.getMessage().contains("SYSTEM"), "Should mention SYSTEM: " + ex.getMessage());
     }
 
 }

--- a/fe/fe-core/src/test/java/com/starrocks/catalog/system/sys/SysFeMemoryUsageTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/catalog/system/sys/SysFeMemoryUsageTest.java
@@ -25,12 +25,16 @@ import com.starrocks.sql.analyzer.Authorizer;
 import com.starrocks.thrift.TAuthInfo;
 import com.starrocks.thrift.TFeMemoryReq;
 import mockit.Expectations;
+import mockit.Mock;
+import mockit.MockUp;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.thrift.TException;
 import org.junit.jupiter.api.Test;
 
 import java.util.Map;
 
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 public class SysFeMemoryUsageTest {
@@ -57,6 +61,29 @@ public class SysFeMemoryUsageTest {
 
         var res = SysFeMemoryUsage.listFeMemoryUsage(req);
         assertTrue(StringUtils.isNotEmpty(res.toString()));
+    }
+
+    @Test
+    public void testListFeMemoryUsageAccessDeniedSurfacesMessage() {
+        TFeMemoryReq req = new TFeMemoryReq();
+        TAuthInfo auth = new TAuthInfo();
+        auth.setUser("nopriv");
+        auth.setUser_ip("127.0.0.1");
+        req.setAuth_info(auth);
+
+        new MockUp<Authorizer>() {
+            @Mock
+            public void checkSystemAction(ConnectContext context, PrivilegeType privilegeType)
+                    throws AccessDeniedException {
+                throw new AccessDeniedException();
+            }
+        };
+
+        TException ex = assertThrows(TException.class, () -> SysFeMemoryUsage.listFeMemoryUsage(req));
+        assertNotNull(ex.getMessage(), "AccessDenied message should not be null");
+        assertTrue(ex.getMessage().contains("Access denied"), "Should match canonical format: " + ex.getMessage());
+        assertTrue(ex.getMessage().contains("OPERATE"), "Should mention OPERATE: " + ex.getMessage());
+        assertTrue(ex.getMessage().contains("SYSTEM"), "Should mention SYSTEM: " + ex.getMessage());
     }
 
 }

--- a/fe/fe-core/src/test/java/com/starrocks/catalog/system/sys/SysFeMemoryUsageTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/catalog/system/sys/SysFeMemoryUsageTest.java
@@ -79,11 +79,11 @@ public class SysFeMemoryUsageTest {
             }
         };
 
-        TException ex = assertThrows(TException.class, () -> SysFeMemoryUsage.listFeMemoryUsage(req));
+        Exception ex = assertThrows(Exception.class, () -> SysFeMemoryUsage.listFeMemoryUsage(req));
         assertNotNull(ex.getMessage(), "AccessDenied message should not be null");
-        assertTrue(ex.getMessage().contains("Access denied"), "Should match canonical format: " + ex.getMessage());
-        assertTrue(ex.getMessage().contains("OPERATE"), "Should mention OPERATE: " + ex.getMessage());
-        assertTrue(ex.getMessage().contains("SYSTEM"), "Should mention SYSTEM: " + ex.getMessage());
+        assertTrue(ex.getMessage().contains(
+                        "Access denied; you need (at least one of) the OPERATE privilege(s) on SYSTEM for this operation."),
+                "Should match canonical format: " + ex.getMessage());
     }
 
 }

